src/otx/algo/common/utils/assigners/hungarian_matcher.py

```diff
@@ -0,0 +1,334 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+"""DETR style Hungarian matcher for bipartite matching."""
+
+from __future__ import annotations
+
+from functools import partial
+
+import torch
+from otx.algo.common.utils.bbox_overlaps import bbox_overlaps
+from otx.algo.common.utils.utils import sample_point
+from scipy.optimize import linear_sum_assignment
+from torch import Tensor, nn
+from torchvision.ops import box_convert
+
+
+def pair_wise_dice_loss(inputs: Tensor, labels: Tensor) -> Tensor:
+    """A pair wise version of the dice loss, see `dice_loss` for usage.
+
+    Args:
+        inputs (Tensor): A tensor representing a mask
+        labels (Tensor): A tensor with the same shape as inputs.
+            Stores the binary classification labels for each element in inputs
+            (0 for the negative class and 1 for the positive class).
+
+    Returns:
+        Tensor: The computed loss between each pairs.
+    """
+    inputs = inputs.sigmoid().flatten(1)
+    numerator = 2 * torch.matmul(inputs, labels.T)
+    # using broadcasting to get a [num_queries, NUM_CLASSES] matrix
+    denominator = inputs.sum(-1)[:, None] + labels.sum(-1)[None, :]
+    return 1 - (numerator + 1) / (denominator + 1)
+
+
+def pair_wise_sigmoid_cross_entropy_loss(inputs: Tensor, labels: Tensor) -> Tensor:
+    """A pair wise version of the cross entropy loss.
+
+    Args:
+        inputs (Tensor): A tensor representing a mask.
+        labels (Tensor): A tensor with the same shape as inputs.
+            Stores the binary classification labels for each element in inputs
+            (0 for the negative class and 1 for the positive class).
+
+    Returns:
+        Tensor: The computed loss between each pairs.
+    """
+    height_and_width = inputs.shape[1]
+
+    criterion = nn.BCEWithLogitsLoss(reduction="none")
+    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs))
+    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs))
+
+    loss_pos = torch.matmul(cross_entropy_loss_pos / height_and_width, labels.T)
+    loss_neg = torch.matmul(cross_entropy_loss_neg / height_and_width, (1 - labels).T)
+    return loss_pos + loss_neg
+
+
+class HungarianMatcher(nn.Module):
```

Comment:
Have you checked that for RTDetr (OD task) there are no changes during loss computation? So, previous detection matcher outputs equal result comparing with this matcher.

Reply:
I tested it and matcher gives the same result:

```python
def test_hungarian_matcher(self, targets, outputs):
        weight_dict = {
            "cost_class": 1.0,
            "cost_bbox": 1.0,
            "cost_giou": 1.0,
        }
        alpha = 0.25
        gamma = 2.0

        matcher = HungarianMatcher(weight_dict, alpha, gamma)
        old_matcher = OldHungarianMatcher(weight_dict, alpha, gamma)

        # Perform the matching
        matches = matcher(outputs, targets)
        old_matches = old_matcher(outputs, targets)
        torch.allclose(matches[0][0], old_matches[0][0])
        torch.allclose(matches[0][1], old_matches[0][1])
```